File: 2024/01.cpp
#include <algorithm>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <numeric>
#include <vector>

int main() {
  std::ifstream input("input");
  std::vector<int> left, right;
  int first, second;
  while (input >> first >> second) {
    left.push_back(first);
    right.push_back(second);
  }
  input.close();

  // Part Two
  const int similarity_score =
      std::accumulate(left.begin(), left.end(), 0, [right](int acc, int n) {
        return acc + std::count(right.begin(), right.end(), n) * n;
      });

  // Part One
  std::sort(left.begin(), left.end());
  std::sort(right.begin(), right.end());
  int distance = 0;
  for (int i = 0; i < left.size(); ++i) {
    distance += std::abs(left[i] - right[i]);
  }

  std::cout << distance << '\n' << similarity_score << std::endl;
}

File: 2024/01.rs
use std::collections::HashMap;

const INPUT: &str = include_str!("input");

fn main() {
    let (mut left, mut right) = (Vec::new(), Vec::new());
    for line in INPUT.lines() {
        let mut numbers = line
            .split_whitespace()
            .map(|n| n.parse().expect("input should be only numbers"));
        left.push(numbers.next().expect("first split"));
        right.push(numbers.next().expect("second split"));
    }
    left.sort_unstable();
    right.sort_unstable();

    println!(
        "{}\n{}",
        distance(&left, &right),
        similarity_score(&left, &right)
    );
}

fn distance(left: &[i32], right: &[i32]) -> u32 {
    left.iter().zip(right).map(|(l, &r)| l.abs_diff(r)).sum()
}

fn similarity_score(left: &[i32], right: &[i32]) -> i32 {
    let mut right_counts = HashMap::new();
    for &num in right {
        *right_counts.entry(num).or_insert(0) += 1;
    }

    left.iter()
        .map(|&n| n * right_counts.get(&n).unwrap_or(&0))
        .sum()
}
